### ISSUE TITLE
Incorrect treatment of structural type

### ISSUE THREAD

**Issue by Glavo** [lampepfl/dotty#4496]
<pre><font color="#3465A4">scala&gt; </font><font color="#C4A000">import</font> scala.reflect.<font color="#75507B">Selectable</font>.reflectiveSelectable 
<font color="#3465A4">scala&gt; </font>(<font color="#75507B">new</font> <font color="#75507B">Object</font> { <font color="#C4A000">val</font> <font color="#06989A">a</font>: <font color="#75507B">Int</font> = <font color="#CC0000">10</font>} : {<font color="#C4A000">val</font> <font color="#06989A">a</font>: <font color="#75507B">Int</font>}).a 
java.lang.<font color="#75507B">NoSuchFieldException:</font> a
	at java.lang.<font color="#75507B">Class</font>.getField(<font color="#75507B">Class</font>.java:<font color="#CC0000">1703</font>)
	at scala.reflect.<font color="#75507B">Selectable$</font>.selectDynamic$extension(<font color="#75507B">Selectable</font>.scala:<font color="#CC0000">7</font>)
	at scala.reflect.<font color="#75507B">Selectable</font>.selectDynamic(<font color="#75507B">Selectable</font>.scala:<font color="#CC0000">4</font>)
	at rs$line$13$.&lt;init&gt;(rs$line$13:<font color="#CC0000">1</font>)
	at rs$line$13$.&lt;clinit&gt;(rs$line$13)
	at rs$line$13.res9Show(rs$line$13)
	at sun.reflect.<font color="#75507B">NativeMethodAccessorImpl</font>.invoke0(<font color="#75507B">Native</font> <font color="#75507B">Method</font>)
	at sun.reflect.<font color="#75507B">NativeMethodAccessorImpl</font>.invoke(<font color="#75507B">NativeMethodAccessorImpl</font>.java:<font color="#CC0000">62</font>)
	at sun.reflect.<font color="#75507B">DelegatingMethodAccessorImpl</font>.invoke(<font color="#75507B">DelegatingMethodAccessorImpl</font>.java:<font color="#CC0000">43</font>)
	at java.lang.reflect.<font color="#75507B">Method</font>.invoke(<font color="#75507B">Method</font>.java:<font color="#CC0000">498</font>)
	at dotty.tools.repl.<font color="#75507B">Rendering</font>.valueOf(<font color="#75507B">Rendering</font>.scala:<font color="#CC0000">58</font>)
	at dotty.tools.repl.<font color="#75507B">Rendering</font>.renderVal(<font color="#75507B">Rendering</font>.scala:<font color="#CC0000">79</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.displayMembers$<font color="#CC0000">3</font>$$anonfun$<font color="#CC0000">3</font>(ReplDriver.scala:<font color="#CC0000">284</font>)
	at scala.collection.<font color="#75507B">TraversableLike</font>.$anonfun$map$<font color="#CC0000">1</font>(TraversableLike.scala:<font color="#CC0000">234</font>)
	at scala.collection.mutable.<font color="#75507B">ResizableArray</font>.foreach(<font color="#75507B">ResizableArray</font>.scala:<font color="#CC0000">59</font>)
	at scala.collection.mutable.<font color="#75507B">ResizableArray</font>.foreach$(<font color="#75507B">ResizableArray</font>.scala:<font color="#CC0000">52</font>)
	at scala.collection.mutable.<font color="#75507B">ArrayBuffer</font>.foreach(<font color="#75507B">ArrayBuffer</font>.scala:<font color="#CC0000">48</font>)
	at scala.collection.<font color="#75507B">TraversableLike</font>.map(<font color="#75507B">TraversableLike</font>.scala:<font color="#CC0000">234</font>)
	at scala.collection.<font color="#75507B">TraversableLike</font>.map$(<font color="#75507B">TraversableLike</font>.scala:<font color="#CC0000">227</font>)
	at scala.collection.<font color="#75507B">AbstractTraversable</font>.map(<font color="#75507B">Traversable</font>.scala:<font color="#CC0000">104</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.displayMembers$<font color="#CC0000">7</font>(ReplDriver.scala:<font color="#CC0000">284</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.displayDefinitions$$anonfun$<font color="#CC0000">3</font>$$anonfun$<font color="#CC0000">2</font>(ReplDriver.scala:<font color="#CC0000">311</font>)
	at scala.<font color="#75507B">Option</font>.map(<font color="#75507B">Option</font>.scala:<font color="#CC0000">146</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.displayDefinitions$$anonfun$<font color="#CC0000">1</font>(ReplDriver.scala:<font color="#CC0000">311</font>)
	at dotty.tools.dotc.core.<font color="#75507B">Periods</font>.atPhase(<font color="#75507B">Periods</font>.scala:<font color="#CC0000">26</font>)
	at dotty.tools.dotc.core.<font color="#75507B">Phases</font>.atPhase(<font color="#75507B">Phases</font>.scala:<font color="#CC0000">36</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.displayDefinitions(<font color="#75507B">ReplDriver</font>.scala:<font color="#CC0000">316</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.compile$$anonfun$<font color="#CC0000">2</font>(ReplDriver.scala:<font color="#CC0000">241</font>)
	at scala.util.<font color="#75507B">Either</font>.fold(<font color="#75507B">Either</font>.scala:<font color="#CC0000">188</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.compile(<font color="#75507B">ReplDriver</font>.scala:<font color="#CC0000">242</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.interpret(<font color="#75507B">ReplDriver</font>.scala:<font color="#CC0000">202</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.loop$<font color="#CC0000">1</font>(ReplDriver.scala:<font color="#CC0000">150</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.runUntilQuit$$anonfun$<font color="#CC0000">1</font>(ReplDriver.scala:<font color="#CC0000">154</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.withRedirectedOutput$$anonfun$<font color="#CC0000">2</font>$$anonfun$<font color="#CC0000">1</font>(ReplDriver.scala:<font color="#CC0000">163</font>)
	at scala.util.<font color="#75507B">DynamicVariable</font>.withValue(<font color="#75507B">DynamicVariable</font>.scala:<font color="#CC0000">58</font>)
	at scala.<font color="#75507B">Console$</font>.withErr(<font color="#75507B">Console</font>.scala:<font color="#CC0000">192</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.withRedirectedOutput$$anonfun$<font color="#CC0000">1</font>(ReplDriver.scala:<font color="#CC0000">163</font>)
	at scala.util.<font color="#75507B">DynamicVariable</font>.withValue(<font color="#75507B">DynamicVariable</font>.scala:<font color="#CC0000">58</font>)
	at scala.<font color="#75507B">Console$</font>.withOut(<font color="#75507B">Console</font>.scala:<font color="#CC0000">163</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.withRedirectedOutput(<font color="#75507B">ReplDriver</font>.scala:<font color="#CC0000">163</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.runUntilQuit(<font color="#75507B">ReplDriver</font>.scala:<font color="#CC0000">154</font>)
	at dotty.tools.repl.<font color="#75507B">Main$</font>.main(<font color="#75507B">Main</font>.scala:<font color="#CC0000">6</font>)
	at dotty.tools.repl.<font color="#75507B">Main</font>.main(<font color="#75507B">Main</font>.scala)

<font color="#3465A4">scala&gt; </font>(<font color="#75507B">new</font> <font color="#75507B">Object</font> { <font color="#C4A000">def</font> <font color="#06989A">a</font>: <font color="#75507B">Int</font> = <font color="#CC0000">10</font>} : {<font color="#C4A000">def</font> <font color="#06989A">a</font>: <font color="#75507B">Int</font>}).a 
java.lang.<font color="#75507B">NoSuchFieldException:</font> a
	at java.lang.<font color="#75507B">Class</font>.getField(<font color="#75507B">Class</font>.java:<font color="#CC0000">1703</font>)
	at scala.reflect.<font color="#75507B">Selectable$</font>.selectDynamic$extension(<font color="#75507B">Selectable</font>.scala:<font color="#CC0000">7</font>)
	at scala.reflect.<font color="#75507B">Selectable</font>.selectDynamic(<font color="#75507B">Selectable</font>.scala:<font color="#CC0000">4</font>)
	at rs$line$14$.&lt;init&gt;(rs$line$14:<font color="#CC0000">1</font>)
	at rs$line$14$.&lt;clinit&gt;(rs$line$14)
	at rs$line$14.res1<font color="#CC0000">0</font>Show(rs$line$14)
	at sun.reflect.<font color="#75507B">NativeMethodAccessorImpl</font>.invoke0(<font color="#75507B">Native</font> <font color="#75507B">Method</font>)
	at sun.reflect.<font color="#75507B">NativeMethodAccessorImpl</font>.invoke(<font color="#75507B">NativeMethodAccessorImpl</font>.java:<font color="#CC0000">62</font>)
	at sun.reflect.<font color="#75507B">DelegatingMethodAccessorImpl</font>.invoke(<font color="#75507B">DelegatingMethodAccessorImpl</font>.java:<font color="#CC0000">43</font>)
	at java.lang.reflect.<font color="#75507B">Method</font>.invoke(<font color="#75507B">Method</font>.java:<font color="#CC0000">498</font>)
	at dotty.tools.repl.<font color="#75507B">Rendering</font>.valueOf(<font color="#75507B">Rendering</font>.scala:<font color="#CC0000">58</font>)
	at dotty.tools.repl.<font color="#75507B">Rendering</font>.renderVal(<font color="#75507B">Rendering</font>.scala:<font color="#CC0000">79</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.displayMembers$<font color="#CC0000">3</font>$$anonfun$<font color="#CC0000">3</font>(ReplDriver.scala:<font color="#CC0000">284</font>)
	at scala.collection.<font color="#75507B">TraversableLike</font>.$anonfun$map$<font color="#CC0000">1</font>(TraversableLike.scala:<font color="#CC0000">234</font>)
	at scala.collection.mutable.<font color="#75507B">ResizableArray</font>.foreach(<font color="#75507B">ResizableArray</font>.scala:<font color="#CC0000">59</font>)
	at scala.collection.mutable.<font color="#75507B">ResizableArray</font>.foreach$(<font color="#75507B">ResizableArray</font>.scala:<font color="#CC0000">52</font>)
	at scala.collection.mutable.<font color="#75507B">ArrayBuffer</font>.foreach(<font color="#75507B">ArrayBuffer</font>.scala:<font color="#CC0000">48</font>)
	at scala.collection.<font color="#75507B">TraversableLike</font>.map(<font color="#75507B">TraversableLike</font>.scala:<font color="#CC0000">234</font>)
	at scala.collection.<font color="#75507B">TraversableLike</font>.map$(<font color="#75507B">TraversableLike</font>.scala:<font color="#CC0000">227</font>)
	at scala.collection.<font color="#75507B">AbstractTraversable</font>.map(<font color="#75507B">Traversable</font>.scala:<font color="#CC0000">104</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.displayMembers$<font color="#CC0000">7</font>(ReplDriver.scala:<font color="#CC0000">284</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.displayDefinitions$$anonfun$<font color="#CC0000">3</font>$$anonfun$<font color="#CC0000">2</font>(ReplDriver.scala:<font color="#CC0000">311</font>)
	at scala.<font color="#75507B">Option</font>.map(<font color="#75507B">Option</font>.scala:<font color="#CC0000">146</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.displayDefinitions$$anonfun$<font color="#CC0000">1</font>(ReplDriver.scala:<font color="#CC0000">311</font>)
	at dotty.tools.dotc.core.<font color="#75507B">Periods</font>.atPhase(<font color="#75507B">Periods</font>.scala:<font color="#CC0000">26</font>)
	at dotty.tools.dotc.core.<font color="#75507B">Phases</font>.atPhase(<font color="#75507B">Phases</font>.scala:<font color="#CC0000">36</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.displayDefinitions(<font color="#75507B">ReplDriver</font>.scala:<font color="#CC0000">316</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.compile$$anonfun$<font color="#CC0000">2</font>(ReplDriver.scala:<font color="#CC0000">241</font>)
	at scala.util.<font color="#75507B">Either</font>.fold(<font color="#75507B">Either</font>.scala:<font color="#CC0000">188</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.compile(<font color="#75507B">ReplDriver</font>.scala:<font color="#CC0000">242</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.interpret(<font color="#75507B">ReplDriver</font>.scala:<font color="#CC0000">202</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.loop$<font color="#CC0000">1</font>(ReplDriver.scala:<font color="#CC0000">150</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.runUntilQuit$$anonfun$<font color="#CC0000">1</font>(ReplDriver.scala:<font color="#CC0000">154</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.withRedirectedOutput$$anonfun$<font color="#CC0000">2</font>$$anonfun$<font color="#CC0000">1</font>(ReplDriver.scala:<font color="#CC0000">163</font>)
	at scala.util.<font color="#75507B">DynamicVariable</font>.withValue(<font color="#75507B">DynamicVariable</font>.scala:<font color="#CC0000">58</font>)
	at scala.<font color="#75507B">Console$</font>.withErr(<font color="#75507B">Console</font>.scala:<font color="#CC0000">192</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.withRedirectedOutput$$anonfun$<font color="#CC0000">1</font>(ReplDriver.scala:<font color="#CC0000">163</font>)
	at scala.util.<font color="#75507B">DynamicVariable</font>.withValue(<font color="#75507B">DynamicVariable</font>.scala:<font color="#CC0000">58</font>)
	at scala.<font color="#75507B">Console$</font>.withOut(<font color="#75507B">Console</font>.scala:<font color="#CC0000">163</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.withRedirectedOutput(<font color="#75507B">ReplDriver</font>.scala:<font color="#CC0000">163</font>)
	at dotty.tools.repl.<font color="#75507B">ReplDriver</font>.runUntilQuit(<font color="#75507B">ReplDriver</font>.scala:<font color="#CC0000">154</font>)
	at dotty.tools.repl.<font color="#75507B">Main$</font>.main(<font color="#75507B">Main</font>.scala:<font color="#CC0000">6</font>)
	at dotty.tools.repl.<font color="#75507B">Main</font>.main(<font color="#75507B">Main</font>.scala)

</pre>

**Comment by allanrenucci** [comment]
Can be reproduced outside of the repl as well:
```scala
import scala.reflect.Selectable.reflectiveSelectable

object Test {
  def main(args: Array[String]): Unit = {
    println((new Object { val a: Int = 10} : {val a: Int}).a) // crash
    println((new Object { def a: Int = 10} : {def a: Int}).a) // crash
  }
}
```

**Comment by Blaisorblade** [comment]
I see two issues:

- `selectDynamic` catches `NoSuchFieldError` (an `IncompatibleClassChangeError`) rather than `NoSuchFieldException` (a `ReflectiveOperationException`). Trivial to fix, and turns

```scala
scala> import scala.reflect.Selectable.reflectiveSelectable; class Foo { val a: Int = 10 }; (new Foo: {val a: Int}).a
// defined class Foo
java.lang.NoSuchFieldException: a
	at java.lang.Class.getField(Class.java:1703)
	at scala.reflect.Selectable$.selectDynamic$extension(Selectable.scala:7)
	at scala.reflect.Selectable.selectDynamic(Selectable.scala:4)
	at rs$line$1$.<init>(rs$line$1:1)
```

into

```scala
scala> import scala.reflect.Selectable.reflectiveSelectable; class Foo { val a: Int = 10 }; (new Foo: {val a: Int}).a
// defined class Foo
val res0: Int = 10
```

- with that fix the original example fails with a different (and confusing) error. Looking at the JDK sources and decompiling the output suggests that's because the generated anonymous class is private, and that's the one introducing the field and accessor. Once I drop the import `scalac` produces _public_ classes. Here instead the `lambdaLift` phase generates private classes.

The access modifier is [fixed to private during `lambdaLift`](https://github.com/lampepfl/dotty/blob/9f110e20ae042ebb9144c8f7ab3382e288afa5b8/compiler/src/dotty/tools/dotc/transform/LambdaLift.scala#L344), and if I crudely drop `| Private` that the original example works (though that's likely not the appropriate solution).

```scala
scala> import scala.reflect.Selectable.reflectiveSelectable; (new Object { val a: Int = 10} : {val a: Int}).a
java.lang.IllegalAccessException: Class scala.reflect.Selectable$ can not access a member of class rs$line$1$$anon$1 with modifiers "public"
	at sun.reflect.Reflection.ensureMemberAccess(Reflection.java:102)
	at java.lang.reflect.AccessibleObject.slowCheckMemberAccess(AccessibleObject.java:296)
	at java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:288)
	at java.lang.reflect.Method.invoke(Method.java:491)
	at scala.reflect.Selectable$.$anonfun$selectDynamicMethod$2(Selectable.scala:22)
	at scala.reflect.Selectable$.selectDynamic$extension(Selectable.scala:12)
	at scala.reflect.Selectable.selectDynamic(Selectable.scala:4)
	at rs$line$1$.<init>(rs$line$1:1)
	at rs$line$1$.<clinit>(rs$line$1)
	at rs$line$1.res0Show(rs$line$1)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

```scala
import scala.reflect.Selectable.reflectiveSelectable; (new Object { val a: Int = 10} : {val a: Int}).a
```

**Comment by Blaisorblade** [comment]
FWIW the JDK code I consulted is http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/sun/reflect/Reflection.java but beware usual license warnings if you want to work on JRE stdlib alternatives.

**Comment by Blaisorblade** [comment]
TLDR:
- Scalac always uses `setAccessible` to bypass access control for reflective access.
- Making anonymous classes public (like Scalac) would reduce non-reflective access control and still not solve the problem in enough cases. Moreover, Dotty expends some effort to make the lifted classes private if possible (see for instance discussion in #517).

## Scalac behavior

The Scala reflection likes to call `setAccessible` through `ensureAccessible`, and calls to it are generated by scalac in
https://github.com/scala/scala/blob/7cb938bb310c9d4af298894c92bac33918179eff/src/compiler/scala/tools/nsc/transform/CleanUp.scala#L92.

By calling `getMethod` (like us), they ensure they only access public methods.
(I suppose they must also call `getField` to also handle Java-defined classes, but we should test this).

## Reflective field access should succeed even on private classes, at least for fields inherited from public classes

Scalac supports reflective calls even on private classes (example courtesy of @allanrenucci), probably by using the `setAccessible` API (this fails instead in Dotty with `IllegalAccessException`):

```scala
scala> import scala.language.reflectiveCalls
import scala.language.reflectiveCalls

scala>  trait Foo { def a: Int }
defined trait Foo

// This code appears clearly well-typed.
// XXX for a realistic use case imagine more indirection between the upcast and the access, as in better-files.
scala> def consumeFoo(foo: Foo) = (foo: { def a: Int }).a
consumeFoo: (foo: Foo)Int

// This code appears also clearly well-typed: Even if Bar is private, it can clearly be upcast to Foo
scala> object A { private class Bar extends Foo { def a = 42 }; def tester = consumeFoo(new Bar) }

scala> A.tester
defined class Bar
res0: Int = 42
```

On the better-files usecase see https://github.com/lampepfl/dotty/issues/1175#issuecomment-242684995.

## Existing decisions in Dotty

@allanrenucci suggested we might not want to enable reflective access to a member first introduced in an anonymous class. What _has_ been clearly decided is that we never infer and export structural types in interfaces (e.g. #1743), but I think this is different, and as shown above Dotty fails even accessing members introduced in accessible superclasses.

**Comment by Blaisorblade** [comment]
Just learned that `{var a: Int}` is legal in Scalac and illegal in Dotty, though its desugaring is legal (but suffers from #4528).
